File: Swift/SCEE.swift
import Foundation

public class SecureCompatibleEncryptionExamples {
    private static let algorithmNonceSize: Int = 12
    private static let algorithmKeySize: Int = 16
    private static let PBKDF2SaltSize: Int = 16
    private static let PBKDF2Iterations: UInt32 = 32767
    
    public static func encryptString(plaintext: String, password: String) throws -> String {
        // Generate a 128-bit salt using a CSPRNG.
        var saltBytes: [UInt8] = [UInt8](repeating: 0, count: PBKDF2SaltSize)
        if SecRandomCopyBytes(kSecRandomDefault, PBKDF2SaltSize, &saltBytes) != errSecSuccess {
            throw SCEEError.CSPRNGFailure
        }
        let salt: Data = Data(bytes: saltBytes)
        
        // Use PBKDF2 to derive a key.
        let passwordData: Data = password.data(using: String.Encoding.utf8)!
        var key: Data = Data(repeating: 0, count: algorithmKeySize)
        
        let status = key.withUnsafeMutableBytes { keyBytes in
            salt.withUnsafeBytes { saltBytes in
                passwordData.withUnsafeBytes { passwordBytes in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordBytes,
                        passwordData.count,
                        saltBytes,
                        PBKDF2SaltSize,
                        CCPBKDFAlgorithm(kCCPRFHmacAlgSHA256),
                        PBKDF2Iterations,
                        keyBytes,
                        algorithmKeySize)
                }
            }
        }
        
        if status != kCCSuccess { throw SCEEError.PBKDF2Failure }
        
        // Encrypt and prepend salt.
        let ciphertextAndNonce: Data = try encrypt(plaintext: plaintext.data(using: String.Encoding.utf8)!, key: key)
        var ciphertextAndNonceAndSalt: Data = Data()
        ciphertextAndNonceAndSalt.append(salt)
        ciphertextAndNonceAndSalt.append(ciphertextAndNonce)
        
        // Return as base64 string.
        return ciphertextAndNonceAndSalt.base64EncodedString()
    }
    
    public static func decryptString(ciphertext: String, password: String) throws -> String {
        // Decode the base64.
        let ciphertextAndNonceAndSalt: Data! = Data(base64Encoded: ciphertext)
        if ciphertextAndNonceAndSalt == nil { throw SCEEError.Base64Failure }
        
        // Retrieve the salt and ciphertextAndNonce.
        let salt: Data = ciphertextAndNonceAndSalt[0..<PBKDF2SaltSize]
        let ciphertextAndNonce: Data = ciphertextAndNonceAndSalt[PBKDF2SaltSize...]
        
        // Use PBKDF2 to derive a key.
        let passwordData: Data = password.data(using: String.Encoding.utf8)!
        var key: Data = Data(repeating: 0, count: algorithmKeySize)
        
        let status = key.withUnsafeMutableBytes { keyBytes in
            salt.withUnsafeBytes { saltBytes in
                passwordData.withUnsafeBytes { passwordBytes in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordBytes,
                        passwordData.count,
                        saltBytes,
                        PBKDF2SaltSize,
                        CCPBKDFAlgorithm(kCCPRFHmacAlgSHA256),
                        PBKDF2Iterations,
                        keyBytes,
                        algorithmKeySize)
                }
            }
        }
        
        if status != kCCSuccess { throw SCEEError.PBKDF2Failure }
        
        // Decrypt and return result.
        return String(data: try decrypt(ciphertextAndNonce: ciphertextAndNonce, key: key), encoding: String.Encoding.utf8)!
    }
    
    public static func encrypt(plaintext: Data, key: Data) throws -> Data {
        // Generate a 96-bit nonce using a CSPRNG.
        var nonceBytes: [UInt8] = [UInt8](repeating: 0, count: algorithmNonceSize)
        if SecRandomCopyBytes(kSecRandomDefault, algorithmNonceSize, &nonceBytes) != errSecSuccess {
            throw SCEEError.CSPRNGFailure
        }
        let nonce: Data = Data(bytes: nonceBytes)
        
        // Create the cipher instance and initialize.
        let gcm: SwiftGCM = try SwiftGCM(key: key, nonce: nonce, tagSize: 16)
        
        // Encrypt and prepend nonce.
        let ciphertext: Data = try gcm.encrypt(auth: nil, plaintext: plaintext)
        var ciphertextAndNonce: Data = Data()
        ciphertextAndNonce.append(nonce)
        ciphertextAndNonce.append(ciphertext)
        
        return ciphertextAndNonce
    }
    
    public static func decrypt(ciphertextAndNonce: Data, key: Data) throws -> Data {
        // Retrieve the nonce and ciphertext.
        let nonce: Data = ciphertextAndNonce[ciphertextAndNonce.startIndex..<ciphertextAndNonce.startIndex + algorithmNonceSize]
        
        let ciphertext: Data = ciphertextAndNonce[ciphertextAndNonce.startIndex + algorithmNonceSize..<ciphertextAndNonce.startIndex + ciphertextAndNonce.count]
        
        // Create the cipher instance and initialize.
        let gcm: SwiftGCM = try SwiftGCM(key: key, nonce: nonce, tagSize: 16)
        
        // Decrypt and return result.
        return try gcm.decrypt(auth: nil, ciphertext: ciphertext)
    }
}

public enum SCEEError: Error {
    case CSPRNGFailure
    case PBKDF2Failure
    case Base64Failure
}

File: Perl/SCEE.pl
use Crypt::AuthEnc::GCM qw(gcm_encrypt_authenticate gcm_decrypt_verify);
use Crypt::PRNG qw(random_bytes);
use Crypt::KeyDerivation qw(pbkdf2);
use MIME::Base64 qw(encode_base64 decode_base64);


use constant ALGORITHM_NONCE_SIZE => 12;
use constant ALGORITHM_TAG_SIZE   => 16;
use constant ALGORITHM_KEY_SIZE   => 16;
use constant PBKDF2_SALT_SIZE     => 16;
use constant PBKDF2_ITERATIONS    => 32767;
use constant PBKDF2_NAME          => "SHA256";


sub encrypt_string {
    my ($plaintext, $password) = @_;

    # Generate a 128-bit salt using a CSPRNG.
    my $salt = random_bytes(PBKDF2_SALT_SIZE);

    # Derive a key.
    my $key = pbkdf2($password, $salt, PBKDF2_ITERATIONS, PBKDF2_NAME, ALGORITHM_KEY_SIZE);

    # Encrypt and prepend salt and return as base64 string.
    return encode_base64($salt . encrypt($plaintext, $key), '');
}

sub decrypt_string {
    my ($base64CiphertextAndNonceAndSalt, $password) = @_;

    # Decode the base64.
    my $ciphertextAndNonceAndSalt = decode_base64($base64CiphertextAndNonceAndSalt);

    # Retrieve the salt and ciphertextAndNonce.
    my ($salt, $ciphertextAndNonce) = unpack("a" . PBKDF2_SALT_SIZE . "a*", $ciphertextAndNonceAndSalt);

    # Derive the key.
    my $key = pbkdf2($password, $salt, PBKDF2_ITERATIONS, PBKDF2_NAME, ALGORITHM_KEY_SIZE);

    # Decrypt and return result.
    return decrypt($ciphertextAndNonce, $key);
}

sub encrypt {
    my ($plaintext, $key) = @_;

    # Generate a 96-bit nonce using a CSPRNG.
    my $nonce = random_bytes(ALGORITHM_NONCE_SIZE);

    # Encrypt and prepend nonce.
    my ($ciphertext, $tag) = gcm_encrypt_authenticate('AES', $key, $nonce, '', $plaintext);
    die "Invalid tag size" unless length $tag == ALGORITHM_TAG_SIZE;
    my $ciphertextAndNonce = $nonce . $ciphertext . $tag;

    return $ciphertextAndNonce;
}

sub decrypt {
    my ($ciphertextAndNonce, $key) = @_;

    # Retrieve the nonce and ciphertext.
    my ($nonce, $ciphertext, $tag) = unpack("a" . ALGORITHM_NONCE_SIZE . 
                                            "a" . (length($ciphertextAndNonce) - ALGORITHM_NONCE_SIZE - ALGORITHM_TAG_SIZE) .
                                            "a" . ALGORITHM_TAG_SIZE, $ciphertextAndNonce);

    # Decrypt and return result.
    my $plaintext = gcm_decrypt_verify('AES', $key, $nonce, '', $ciphertext, $tag);

    return $plaintext;
}


1;
